docs/content/guides/dagster/model_selection.mdx

```diff
@@ -0,0 +1,231 @@
+---
+title: Model Selection | Dagster
+description: This guide describes how to use Dagster for exploratory model selection for machine learning.
+---
+
+# Model Selection
+
+Every machine learning problem begins with exploration. Before training a final model or building an online learning system, there are many decisions to make. One of the most important decisions is: _what learning algorithm should be used, and how should it be configured_? This is known as _model selection._
+
+There are many learning algorithms for common tasks (e.g. classification, regression). Each algorithm defines a set of configurable hyperparameters, and each hyperparameter may range over a potentially infinite set of values. We can call a learning algorithm together with a fixed set of hyperparameter bindings a _model architecture_. Model selection then involves choosing a single model architecture out of an infinite set of possibilities.
```

Comment:
This might be a pedantic distinction, but a couple concerns with framing this as model architecture selection:
* I think many people understand the term "model architecture" to mean a learning algorithm _without_ the hyperparameters yet specified. E.g. one might say "we used a two-level CNN architecture, and we set the learning rate to 0.9".
* I suspect that, in many cases, the goal is model selection, not architecture/hyperparameter selection.  E.g. every week you want to train an ML model.  When you do that, you consider a bunch of architectures/hyperparameters.  But the next week, you don't just stick with the architecture/hyperparameters that worked best in the previous week - you again try training a bunch of models with a bunch of architectures/hyperparameters and again use the best one.

Reply:
I don't think these points are pedantic-- I think it's important to be very precise when talking about machine learning. IMO there's enormous confusion among people because not enough authors have been pedantic enough! 

This is framed as "model architecture selection" bc this is the most accurate description I came up with of what we are actually doing. Here's why:

As you correctly point out, our ultimate goal is **model selection**. However, the naive approach to this of "training a bunch of models with a bunch of architectures/hyperparameters and [using] the best one" is biased because it overfits the training data. You can't get away from this problem, and so you shouldn't select model with this straightforward method.

Instead, *the rigorous way to do model selection is via model architecture selection.* You train a bunch of models, evaluate the performance of these models as a means of selecting the best *model architecture*, and then train a single final model (using the optimal architecture) on your entire training set (none of the models used to evaluate candidate architectures are trained on the whole training set). That final model is your answer to the model selection problem.

Also ,I'm open to using a different term than "model architecture", and I agree that it is somewhat ambiguous-- unfortunately there does not seem to be a better available term of art for this important concept: a fully-configured (i.e. all hyperparameters specified) learning algorithm. Part of the ambiguity arises from the fact that  hyperparameters often (e.g. number of hidden units in a neural network) map directly to "spatial", i.e. architectural, characteristics of the algorithm (unlike learning rate). In any case, this very ambiguity is part of the reason the background is as long as it is. With no unambiguous dominant lexicon, I have to set up clear meanings for these terms within the article scope.